File: dart/lib/store/metadata.dart
import 'dart:convert';
import 'package:sqlite3/sqlite3.dart';
import 'package:uuid/uuid.dart';
import 'scope.dart';
import 'types.dart';

const _schema = '''
CREATE TABLE IF NOT EXISTS knowledge (
  id TEXT PRIMARY KEY,
  content TEXT NOT NULL,
  type TEXT NOT NULL CHECK(type IN ('pattern','preference','decision','mistake','workaround')),
  scope TEXT NOT NULL CHECK(scope IN ('global','project')),
  project TEXT,
  tags TEXT NOT NULL DEFAULT '[]',
  session_id TEXT NOT NULL,
  "trigger" TEXT NOT NULL CHECK("trigger" IN ('pre_compact','session_end','manual')),
  source_timestamp TEXT NOT NULL,
  confidence REAL NOT NULL DEFAULT 0.5,
  access_count INTEGER NOT NULL DEFAULT 0,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL
);

CREATE INDEX IF NOT EXISTS idx_knowledge_scope ON knowledge(scope);
CREATE INDEX IF NOT EXISTS idx_knowledge_type ON knowledge(type);
CREATE INDEX IF NOT EXISTS idx_knowledge_project ON knowledge(project);
''';

const _uuid = Uuid();

/// Metadata store backed by SQLite.
class MetadataStore {
  final Database _db;

  MetadataStore(KnowledgeScope scope, {String? projectRoot})
      : _db = sqlite3.open(resolveDbPath(scope, projectRoot: projectRoot)) {
    _db.execute('PRAGMA journal_mode = WAL');
    _db.execute(_schema);
  }

  /// Insert a new knowledge chunk, returns full chunk with generated id/timestamps.
  KnowledgeChunk insert(KnowledgeInput input) {
    final now = DateTime.now().toUtc().toIso8601String();
    final id = _uuid.v4();

    final stmt = _db.prepare('''
      INSERT INTO knowledge (id, content, type, scope, project, tags, session_id, "trigger", source_timestamp, confidence, access_count, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?)
    ''');

    stmt.execute([
      id,
      input.content,
      input.type.name,
      input.scope.name,
      input.project,
      jsonEncode(input.tags),
      input.source.sessionId,
      input.source.trigger.value,
      input.source.timestamp,
      input.confidence,
      now,
      now,
    ]);
    stmt.dispose();

    return KnowledgeChunk(
      id: id,
      content: input.content,
      type: input.type,
      scope: input.scope,
      project: input.project,
      tags: List<String>.from(input.tags),
      source: input.source,
      confidence: input.confidence,
      accessCount: 0,
      createdAt: now,
      updatedAt: now,
    );
  }

  /// Get a knowledge chunk by ID.
  KnowledgeChunk? getById(String id) {
    final result = _db.select('SELECT * FROM knowledge WHERE id = ?', [id]);
    if (result.isEmpty) return null;
    return _rowToChunk(result.first);
  }

  /// Search by filters (non-vector, metadata only).
  List<KnowledgeChunk> search({
    KnowledgeScope? scope,
    KnowledgeType? type,
    String? project,
    int limit = 20,
  }) {
    final conditions = <String>[];
    final params = <Object?>[];

    if (scope != null) {
      conditions.add('scope = ?');
      params.add(scope.name);
    }
    if (type != null) {
      conditions.add('type = ?');
      params.add(type.name);
    }
    if (project != null) {
      conditions.add('project = ?');
      params.add(project);
    }

    final where =
        conditions.isNotEmpty ? 'WHERE ${conditions.join(' AND ')}' : '';

    params.add(limit);

    final result = _db.select(
      'SELECT * FROM knowledge $where ORDER BY updated_at DESC LIMIT ?',
      params,
    );

    return result.map(_rowToChunk).toList();
  }

  /// Increment access count (called on recall).
  void touch(String id) {
    final stmt = _db.prepare(
      'UPDATE knowledge SET access_count = access_count + 1, updated_at = ? WHERE id = ?',
    );
    stmt.execute([DateTime.now().toUtc().toIso8601String(), id]);
    stmt.dispose();
  }

  /// Update scope (promote/demote).
  void updateScope(String id, KnowledgeScope newScope) {
    final stmt = _db.prepare(
      'UPDATE knowledge SET scope = ?, updated_at = ? WHERE id = ?',
    );
    stmt.execute([newScope.name, DateTime.now().toUtc().toIso8601String(), id]);
    stmt.dispose();
  }

  /// Delete a knowledge entry.
  bool delete(String id) {
    final stmt = _db.prepare('DELETE FROM knowledge WHERE id = ?');
    stmt.execute([id]);
    stmt.dispose();
    return _db.updatedRows > 0;
  }

  /// Get aggregate statistics.
  MetadataStats stats() {
    final totalResult =
        _db.select('SELECT COUNT(*) as cnt FROM knowledge');
    final total = totalResult.first['cnt'] as int;

    final byType = <String, int>{};
    final typeRows =
        _db.select('SELECT type, COUNT(*) as cnt FROM knowledge GROUP BY type');
    for (final row in typeRows) {
      byType[row['type'] as String] = row['cnt'] as int;
    }

    final byScope = <String, int>{};
    final scopeRows = _db
        .select('SELECT scope, COUNT(*) as cnt FROM knowledge GROUP BY scope');
    for (final row in scopeRows) {
      byScope[row['scope'] as String] = row['cnt'] as int;
    }

    return MetadataStats(total: total, byType: byType, byScope: byScope);
  }

  /// Close the database connection.
  void close() {
    _db.dispose();
  }
}

/// Aggregate statistics for knowledge store.
class MetadataStats {
  final int total;
  final Map<String, int> byType;
  final Map<String, int> byScope;

  const MetadataStats({
    required this.total,
    required this.byType,
    required this.byScope,
  });
}

KnowledgeChunk _rowToChunk(Row row) {
  return KnowledgeChunk(
    id: row['id'] as String,
    content: row['content'] as String,
    type: KnowledgeType.fromString(row['type'] as String),
    scope: KnowledgeScope.fromString(row['scope'] as String),
    project: row['project'] as String?,
    tags: List<String>.from(jsonDecode(row['tags'] as String) as List),
    source: KnowledgeSource(
      sessionId: row['session_id'] as String,
      timestamp: row['source_timestamp'] as String,
      trigger: ExtractionTrigger.fromString(row['trigger'] as String),
    ),
    confidence: (row['confidence'] as num).toDouble(),
    accessCount: row['access_count'] as int,
    createdAt: row['created_at'] as String,
    updatedAt: row['updated_at'] as String,
  );
}

File: python/src/store/metadata.py
"""Metadata store backed by SQLite."""

import json
import sqlite3
import uuid
from datetime import datetime, timezone

from .scope import resolve_db_path
from .types import (
    KnowledgeChunk,
    KnowledgeInput,
    KnowledgeScope,
    KnowledgeSource,
    KnowledgeType,
)

SCHEMA = """
CREATE TABLE IF NOT EXISTS knowledge (
  id TEXT PRIMARY KEY,
  content TEXT NOT NULL,
  type TEXT NOT NULL CHECK(type IN ('pattern','preference','decision','mistake','workaround')),
  scope TEXT NOT NULL CHECK(scope IN ('global','project')),
  project TEXT,
  tags TEXT NOT NULL DEFAULT '[]',
  session_id TEXT NOT NULL,
  trigger TEXT NOT NULL CHECK("trigger" IN ('pre_compact','session_end','manual')),
  source_timestamp TEXT NOT NULL,
  confidence REAL NOT NULL DEFAULT 0.5,
  access_count INTEGER NOT NULL DEFAULT 0,
  created_at TEXT NOT NULL,
  updated_at TEXT NOT NULL
);

CREATE INDEX IF NOT EXISTS idx_knowledge_scope ON knowledge(scope);
CREATE INDEX IF NOT EXISTS idx_knowledge_type ON knowledge(type);
CREATE INDEX IF NOT EXISTS idx_knowledge_project ON knowledge(project);
"""


class MetadataStore:
    """Metadata store backed by SQLite."""

    def __init__(self, scope: KnowledgeScope, project_root: str | None = None) -> None:
        db_path = resolve_db_path(scope, project_root)
        self.db = sqlite3.connect(db_path)
        self.db.row_factory = sqlite3.Row
        self.db.execute("PRAGMA journal_mode = WAL")
        self.db.executescript(SCHEMA)

    def insert(self, input_: KnowledgeInput) -> KnowledgeChunk:
        """Insert a new knowledge chunk, returns full chunk with generated id/timestamps."""
        now = datetime.now(timezone.utc).isoformat()
        chunk_id = str(uuid.uuid4())

        self.db.execute(
            """
            INSERT INTO knowledge (id, content, type, scope, project, tags, session_id, "trigger", source_timestamp, confidence, access_count, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?)
            """,
            (
                chunk_id,
                input_.content,
                input_.type,
                input_.scope,
                input_.project,
                json.dumps(input_.tags),
                input_.source.session_id,
                input_.source.trigger,
                input_.source.timestamp,
                input_.confidence,
                now,
                now,
            ),
        )
        self.db.commit()

        return KnowledgeChunk(
            id=chunk_id,
            content=input_.content,
            type=input_.type,
            scope=input_.scope,
            project=input_.project,
            tags=input_.tags,
            source=input_.source,
            confidence=input_.confidence,
            access_count=0,
            created_at=now,
            updated_at=now,
        )

    def get_by_id(self, chunk_id: str) -> KnowledgeChunk | None:
        """Get a knowledge chunk by ID."""
        cursor = self.db.execute("SELECT * FROM knowledge WHERE id = ?", (chunk_id,))
        row = cursor.fetchone()
        return _row_to_chunk(row) if row else None

    def search(
        self,
        scope: KnowledgeScope | None = None,
        type_: KnowledgeType | None = None,
        project: str | None = None,
        limit: int = 20,
    ) -> list[KnowledgeChunk]:
        """Search by filters (non-vector, metadata only)."""
        conditions: list[str] = []
        params: list[object] = []

        if scope:
            conditions.append("scope = ?")
            params.append(scope)
        if type_:
            conditions.append("type = ?")
            params.append(type_)
        if project:
            conditions.append("project = ?")
            params.append(project)

        where = f"WHERE {' AND '.join(conditions)}" if conditions else ""
        params.append(limit)

        cursor = self.db.execute(
            f"SELECT * FROM knowledge {where} ORDER BY updated_at DESC LIMIT ?",
            params,
        )
        rows = cursor.fetchall()
        return [_row_to_chunk(row) for row in rows]

    def touch(self, chunk_id: str) -> None:
        """Increment access count (called on recall)."""
        now = datetime.now(timezone.utc).isoformat()
        self.db.execute(
            "UPDATE knowledge SET access_count = access_count + 1, updated_at = ? WHERE id = ?",
            (now, chunk_id),
        )
        self.db.commit()

    def update_scope(self, chunk_id: str, new_scope: KnowledgeScope) -> None:
        """Update scope (promote/demote)."""
        now = datetime.now(timezone.utc).isoformat()
        self.db.execute(
            "UPDATE knowledge SET scope = ?, updated_at = ? WHERE id = ?",
            (new_scope, now, chunk_id),
        )
        self.db.commit()

    def delete(self, chunk_id: str) -> bool:
        """Delete a knowledge entry."""
        cursor = self.db.execute("DELETE FROM knowledge WHERE id = ?", (chunk_id,))
        self.db.commit()
        return cursor.rowcount > 0

    def stats(self) -> dict:
        """Get aggregate statistics."""
        total_row = self.db.execute("SELECT COUNT(*) as cnt FROM knowledge").fetchone()
        total = total_row["cnt"]

        by_type: dict[str, int] = {}
        for row in self.db.execute(
            "SELECT type, COUNT(*) as cnt FROM knowledge GROUP BY type"
        ):
            by_type[row["type"]] = row["cnt"]

        by_scope: dict[str, int] = {}
        for row in self.db.execute(
            "SELECT scope, COUNT(*) as cnt FROM knowledge GROUP BY scope"
        ):
            by_scope[row["scope"]] = row["cnt"]

        return {"total": total, "byType": by_type, "byScope": by_scope}

    def close(self) -> None:
        """Close the database connection."""
        self.db.close()


def _row_to_chunk(row: sqlite3.Row) -> KnowledgeChunk:
    return KnowledgeChunk(
        id=row["id"],
        content=row["content"],
        type=row["type"],
        scope=row["scope"],
        project=row["project"],
        tags=json.loads(row["tags"]),
        source=KnowledgeSource(
            session_id=row["session_id"],
            timestamp=row["source_timestamp"],
            trigger=row["trigger"],
        ),
        confidence=row["confidence"],
        access_count=row["access_count"],
        created_at=row["created_at"],
        updated_at=row["updated_at"],
    )
